specification/dartLangSpec.tex

```diff
@@ -16537,62 +16542,69 @@ \subsection{Identifier Reference}
+but we specify its syntax here because it is used in several different contexts,
+and it is more closely related to the plain identifier
+than it is to any single one of those grammar rules where it is used.%
+}
 
 \begin{grammar}
 <identifier> ::= <IDENTIFIER>
+  \alt <BUILT\_IN\_IDENTIFIER>
+  \alt <OTHER\_IDENTIFIER>
 
-<IDENTIFIER\_NO\_DOLLAR> ::= <IDENTIFIER\_START\_NO\_DOLLAR>
-  \gnewline{} <IDENTIFIER\_PART\_NO\_DOLLAR>*
+<typeIdentifier> ::= <IDENTIFIER>
+  \alt <OTHER\_IDENTIFIER>
 
-<IDENTIFIER> ::= <IDENTIFIER\_START> <IDENTIFIER\_PART>*
+<qualifiedName> ::= <typeIdentifier> `.' <identifier>
+  \alt <typeIdentifier> `.' <typeIdentifier> `.' <identifier>
 
-<BUILT\_IN\_IDENTIFIER> ::= \ABSTRACT{}
-  \alt \AS{}
-  \alt \COVARIANT{}
-  \alt \DEFERRED{}
-  \alt \DYNAMIC{}
-  \alt \EXPORT{}
-  \alt \EXTERNAL{}
-  \alt \EXTENSION{}
-  \alt \FACTORY{}
-  \alt \FUNCTION{}
-  \alt \GET{}
-  \alt \IMPLEMENTS{}
-  \alt \IMPORT{}
-  \alt \INTERFACE{}
-  \alt \LATE{}
-  \alt \LIBRARY{}
-  \alt \MIXIN{}
-  \alt \OPERATOR{}
-  \alt \PART{}
-  \alt \REQUIRED{}
-  \alt \SET{}
-  \alt \STATIC{}
-  \alt \TYPEDEF{}
+<BUILT\_IN\_IDENTIFIER> ::= \ABSTRACT{} | \AS{} | \COVARIANT{} | \DEFERRED{}
+  \alt\hspace{-3mm} \DYNAMIC{} | \EXPORT{} | \EXTERNAL{} | \EXTENSION{} |
+    \FACTORY{} | \FUNCTION{} | \GET{}
+  \alt\hspace{-3mm} \IMPLEMENTS{} | \IMPORT{} | \INTERFACE{} | \LATE{} |
+    \LIBRARY{} | \MIXIN{} | \OPERATOR{}
+  \alt\hspace{-3mm} \PART{} | \REQUIRED{} | \SET{} | \STATIC{} | \TYPEDEF{}
 
-<IDENTIFIER\_START> ::= <IDENTIFIER\_START\_NO\_DOLLAR> | `$'
+<OTHER\_IDENTIFIER> ::= \gnewline{}
+  \ASYNC{} | \HIDE{} | \OF{} | \ON{} | \SHOW{} | \SYNC{} |
+  \AWAIT{} | \YIELD{}
+
+<IDENTIFIER\_NO\_DOLLAR> ::= <IDENTIFIER\_START\_NO\_DOLLAR>
+  \gnewline{} <IDENTIFIER\_PART\_NO\_DOLLAR>*
 
 <IDENTIFIER\_START\_NO\_DOLLAR> ::= <LETTER> | `_'
 
 <IDENTIFIER\_PART\_NO\_DOLLAR> ::= \gnewline{}
   <IDENTIFIER\_START\_NO\_DOLLAR> | <DIGIT>
 
-<IDENTIFIER\_PART> ::= <IDENTIFIER\_START> | <DIGIT>
+<IDENTIFIER> ::= <IDENTIFIER\_START> <IDENTIFIER\_PART>*
 
-<qualifiedName> ::= <typeIdentifier> `.' <identifier>
-  \alt <typeIdentifier> `.' <typeIdentifier> `.' <identifier>
+<IDENTIFIER\_START> ::= <IDENTIFIER\_START\_NO\_DOLLAR> | `$'
+
+<IDENTIFIER\_PART> ::= <IDENTIFIER\_START> | <DIGIT>
 
-<LETTER> ::= `a' .. `z'
-  \alt `A' .. `Z'
+<LETTER> ::= `a' .. `z' | `A' .. `Z'
 
 <DIGIT> ::= `0' .. `9'
 
 <WHITESPACE> ::= (`\\t' | ` ' | <NEWLINE>)+
 \end{grammar}
 
 \LMHash{}%
-A built-in identifier is one of the identifiers produced by the production
-\synt{BUILT\_IN\_IDENTIFIER}.
+The ordering of the lexical rules above ensure that \synt{IDENTIFIER}
+and \synt{IDENTIFIER\_NO\_DOLLAR} do not derive any built-in identifiers.
+Similarly, the lexical rule for reserved words (\ref{reservedWords})
+must be considered to come before the rule for \synt{BUILT\_IN\_IDENTIFIER},
+such that \synt{IDENTIFIER} and \synt{IDENTIFIER\_NO\_DOLLAR}
+also do not derive any reserved words.
```

Comment:
Hmm.

When we have a grammar rule like 
```ebnf
<statement> ::= ...
   | `for' ...
```
does that `for` not conflict with `<RESERVED_WORD>`? Or can the grammar unpack the `<RESERVED_WORD>` that contains the `for`?

Is that because lexical rules are different from grammatical rules?

(Or can we just do:
```ebnf
<IDENTIFIER> ::= 
   NEGATIVE_LOOKAHEAD(<RESERVED_WORD>|<BUILT_IN_IDENTIFIER>)
   <IDENTIFIER\_START> <IDENTIFIER\_PART>*
```
instead, and not rely on ordering?

Reply:
It is completely standard that a language has certain reserved words, and identifiers are defined using a simple regular expression. In practice this is achieved by giving the lexical rule for identifiers after the rules for reserved words. I would expect most language specification documents to rely on this approach implicitly. It may seem somewhat hacky to rely on the ordering, but the real difference would simply be that we're admitting it.

ANTLR requires lexical productions for each word (so we have `FOR ::= 'for'` and such), but the grammar in the language specification uses the keywords directly. I believe this is a reasonable compromise: We rely on the spec grammar to identify multiple occurrences of words. However, ANTLR already treats sequences of punctuation characters in that manner (so there's no need to define `PLUSPLUS ::= '++'` even though '++' occurs more than once in the grammar).

On the other hand, I believe that negative lookahead is considerably more esoteric than relying on the ordering of the lexical rules.

I'm keeping the rules as they are for now, please create an issue to get it changed if needed.